vermod.8
.\" section 8 man page for vermod
.\"
.\" This man page was first written by Cody Boone Ferguson for the IOCCC
.\" in 2022.
.\"
.\" Humour impairment is not virtue nor is it a vice, it's just plain
.\" wrong: almost as wrong as JSON spec misfeatures and C++ obfuscation! :-)
.\"
.\" "Share and Enjoy!"
.\"     --  Sirius Cybernetics Corporation Complaints Division, JSON spec department. :-)
.\"
.TH vermod 8 "22 October 2022" "vermod" "IOCCC tools"
.SH NAME
vermod \- modify version strings (and others) under ./test_JSON/
.SH SYNOPSIS
\fBvermod.sh\fP [\-h] [\-V] [\-v level] [\-d test_dir] [\-i limit.sh] [\-o] [\-F] [\-Q] [\-n] [\-l] [\-L] old_ver new_ver
.SH DESCRIPTION
\fBvermod.sh\fP uses \fBrpl(1)\fP to update strings from an old value to a new value under ./test_JSON (by default).
This is needed for when a version of a tool changes so that the test suite does not fail.
.SH OPTIONS
.TP
\fB\-h\fP
Show help and exit.
.TP
\fB\-V\fP
Show version and exit.
.TP
\fB\-v\fP
Set verbosity level (def: 0).
.TP
\fB\-d\fP
Set the directory where the '*.json' files are expected.
.TP
\fB\-i\fP
Set limit.sh file to verify new version.
.TP
\fB\-o\fP
Verify old version instead of the new version (which is the default).
.TP
\fB\-F\fP
Force change even if verification fails (def: reject unverified changes).
.TP
\fB\-Q\fP
Do not double quote strings.
.TP
\fB\-n\fP
No op: change no files.
.TP
\fB\-l\fP
List files that change or would be changed.
This is useful with \fB\-n\fP to show what would change.
.TP
\fB\-L\fP
List files that do not change or would not be changed.
This is useful with \fB\-n\fP to show what wouldn't change.
.SH EXIT STATUS
.TP
0
all is well
.TQ
1
\fBrpl\fP exited non\-zero
.TQ
2
\fBrpl\fP not found
.TQ
3
no limit.sh found or is not readable
.TQ
4
no test directory found or is not readable
.TQ
5
usage message due to \-h
.TQ
6
command line error and usage message printed
.TQ
7
no *.json files found under test directory
.TQ
8
new_ver (or old_ver if \-o) not found in limit file
.TQ
>= 10
internal error
.SH FILES
\fItest_JSON\fP
.RS
Default directory of good and bad subdirectories each respectively with good and bad JSON files.
.RE
.SH NOTES
.PP
This script is only needed when a version of a tool is changed by a developer of the repo.
.PP
The tool \fBsed(1)\fP and various others could also be used to do this but this script is generally preferred because it can verify that the version change is actually correct.
.SH EXAMPLES
.PP
.nf
After changing the version of \fBtxzchk\fP to \fB0.12 2022\-09\-2\fP from \fB0.11 2022\-08\-23\fP show what files would change:
.RS
\fB
 ./vermod.sh \-l \-n '0.11 2022\-08\-23' '0.12 2022\-09\-22'\fP
.RE
.PP
.nf
After verifying that the above command is correct, run the script again to actually make the changes:
.RS
\fB
 ./vermod.sh  '0.11 2022\-08\-23' '0.12 2022\-09\-22'\fP
.RE
.SH SEE ALSO
\fBrpl(1)\fP


CHANGES.md
# Major changes to the IOCCC entry toolkit


## Release 0.6 2022-09-02

Released IOCCC entry toolkit v0.6 2022-09-02

Updated CHANGES.md for v0.6 2022-09-02.

Changed `MIN_TIMESTAMP` from 1655958810 to 1662145368.

Updated `mkiocccentry` from "0.40 2022-03-15" to "0.41 2022-09-02".
Updated `.info.json` version from "1.10 2022-06-22" to "1.11 2022-09-02".
Updated `.author.json` version from "1.13 2022-06-22" to "1.14 2022-09-02".

Improved code to use new facilities for output to a buffer from dbg release of
v2.5 2022-07-23.

The `chk_foo()` functions in `chk_validate.c` and the `test_foo()` functions in
`entry_util.c` are code complete, although they remain untested and unused.  The
`chkentry` tool, however, is not yet code complete. Later releases of tested
JSON semantic code will no doubt modify these functions.

Improved a number of the ways that JSON field values are checked.  In a number
of cases, code form `mkiocccentry.c` was moved into `test_foo()` functions so
that they could be used by other tools such as the JSON semantic test code.

Added code to generate JSON semantic tables from JSON reference files for
`.info.json` and `.author.json`.  The `jsemcgen.sh` tool manages this by way of
the `jsemtblgen` code generator and header, patch and trailer files (see
`chk.auth.*` and `chk.info.*` files).

Avoided the appearance of attacking any particular individual.  It was not our
intention to disrespect anyone, even though we disagree with some of the
technical decisions.  Where we have fundamental technical disagreements, we
attempted to express those technical disagreements with humour in hopefully a
more fun way.  As also now apologize for how `bison` and `flex` generated code
may look, instead of simply calling it ugly.  As such, we hopefully improved
some of the humour in our code while trying to be nice and friendly to others.

For example now the adjusted dbg levels in JSON parser are:

```
*     At -J 3, only the top level return type and top level tree are printed.
*     At -J 5, intermediate tree return types and tree are printed too.
*     At -J 7, also print grammar progress.
*     At -J 9, also print sorry_text and sorry_lang grammar values.
```

Removed a number of files and added a number of files under the `test_JSON/`
tree.  When the JSON semantic code is being tested in a future release, we
expect more such `test_JSON/` tree changes.

Improved / add a number of man pages.  Updated `README.md`.

Improved and expanded `txzchk`.

Added more test code.  We attempt to detect feathers in tarballs.  :-)

We will neither confirm nor deny the presence of an "Easter egg".
To to so would be "foolish".  :-)

Improved and fixed `vermod.sh` and `reset_tstamp.sh`.  Tested this
code by changing the `MIN_TIMESTAMP` as noted above.  The `MIN_TIMESTAMP`
needed to up updated anyway due to changes in the `.info.json` and
`.author.json` formats.

Made numerous improvements and bug fixes to the Makefile.

Fixed how `picky` is used in by the `make picky` rule for a few special files.

Added multiple rules to the Makefile including but not limited to `make mkref`,
`make legacy_clobber`, and `make legacy_clean` rules.  Applied multiple bug
fixes to the Makefile.

Improved the Makefile to be less impacting for modern systems while trying to
maintain, for as long as we can, compatibility with some older systems.

Attempted to improve compatibility with reasonably modern systems.  We managed
to keep CentOS 7 somewhat supported for now, although we may be forced to drop
support of such an old system before 2024. But if we do reach 2024 we can only
say that given that CentOS 7 will reach EOL in 2024 and CentOS 8 support was
unduly dropped and since < CentOS 7 already was at EOL this should not be a
problem. :-)

Users of very out of date systems can still enter and submit entries to the
IOCCC.  They just might need to find a more modern system to package and submit
their IOCCC entry, however.

Added stub code for `hostchk.sh`.  A future release will include mode tests for
the given hosts.  Future releases will also include a bug report system that
will also use `hostchk.sh`.

Improved the `no-comment` directive in `.info.json` and `.author.json` files.

Improved how `time_t` values are used and printed.  We no longer assume that
`time_t` is signed nor assume it is unsigned: we only assume it's time. :-)

Improved comments in C code about special `seqcexit` comment directives.

Make numerous bug fixes and fixed a fair number of typos.  Nevertheless we claim
this is *NOT* complete. :-)


## Release 0.5 2022-07-10

The `jsemtblgen` tool is code complete.

Added `is_reserved()` function to `rule_count.c`.

Changed the Official iocccsize version from 28.10 2022-03-15 to 28.11
2022-07-10. Updated `test_JSON` tree and `json_teststr.txt` to reflect the new
`IOCCCSIZE_VERSION` value.

Added and improved a number of man pages.

Added `Makefile` rule to build HTML files from all man pages.

Both the `parse_json()` and `parse_json_file()` top level functions of the JSON
parser return a JSON node parse tree when the JSON is valid.

Trimmed the JSON parser to not include code that does not strictly belong to the
generic JSON parser.  Removed dead C code.

Updated dbg facility to version 2.4 2022-06-26.  Version 2.4 2022-06-26 contains
functions to write (form) a message, debug diagnostic message, warning, or error
message into a string buffer:

```c
extern void snmsg(char *str, size_t size, char const *fmt, ...)
extern void sndbg(char *str, size_t size, int level, char const *fmt, ...);
extern void snwarn(char *str, size_t size, char const *name, char const *fmt, ...);
extern void snwarnp(char *str, size_t size, char const *name, char const *fmt, ...);
extern void snwerr(int error_code, char *str, size_t size, char const *name, char const *fmt, ...);
extern void snwerrp(int error_code, char *str, size_t size, char const *name, char const *fmt, ...);
```


vermod.sh
#!/usr/bin/env bash
#
# vermod.sh - modify version strings (and others) under ./test_JSON
#
# Copyright (c) 2022 by Landon Curt Noll.  All Rights Reserved.
#
# Permission to use, copy, modify, and distribute this software and
# its documentation for any purpose and without fee is hereby granted,
# provided that the above copyright, this permission notice and text
# this comment, and the disclaimer below appear in all of the following:
#
#       supporting documentation
#       source copies
#       source works derived from this source
#       binaries derived from this source or from derived source
#
# LANDON CURT NOLL DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS SOFTWARE,
# INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS. IN NO
# EVENT SHALL LANDON CURT NOLL BE LIABLE FOR ANY SPECIAL, INDIRECT OR
# CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS OF
# USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
# OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
# PERFORMANCE OF THIS SOFTWARE.
#
# chongo (Landon Curt Noll, http://www.isthe.com/chongo/index.html) /\oo/\
#
# Share and enjoy! :-)

# setup
#
export VERMOD_VERSION="0.1 2022-03-15"
export USAGE="usage: $0 [-h] [-V] [-v level] [-d test_dir] [-i limit.sh]
	[-o] [-F] [-Q] [-n] [-l] [-L] old_ver new_ver

    -h		    Print help and exit
    -V		    Print version and exit
    -v level	    Set the debugging level (def: 0)

    -d test_dir	    Process '*.json' files below dir (def: ./test_JSON)
    -i limit.sh	    Use limit.sh to verify new version (def: ./limit_ioccc.sh)

    -o		    Verify old version (def: verify new version)
    -F		    Force change, even if it cannot be verified (def: reject an unverified change)

    -Q		    Do not double quote string (def: add double quotes around old_ver and new_ver when making changes)

    -n		    No op: change no files (def: change files)
    -l		    List files that change/will change (def: do not list)
			NOTE: Useful with -n to show what would change
    -L		    List files that do not/will not change (def: do not list)
			NOTE: Useful with -n to show what will not change

    old_ver	    Old version string to change from
    new_ver	    New version string to change to

Exit codes:
     0   all is well
     1   rpl exited non-zero
     2   rpl command not found
     3   no limit.sh file found, or is not readable
     4   no test_dir directory found, or is not readable
     5   usage message printed due to -h
     6   command line error and usage message printed
     7   no *.json files found under test_dir
     8   new_ver (or old_ver if -o) not found in limit.sh
 >= 10   internal error

$0 version: $VERMOD_VERSION"
export V_FLAG="0"
export JSON_TREE="./test_JSON"
export LIMIT_SH="./limit_ioccc.sh"
export OLD_FLAG=
export FORCE=
export QUOTE='"'
export NOOP=
export LIST_CHANGE=
export LIST_NOCHANGE=

# parse args
#
while getopts :hVv:d:i:oFQnlL flag; do
    case "$flag" in
    h) echo "$USAGE" 1>&2
       exit 5
       ;;
    V) echo "$VERMOD_VERSION" 1>&2
       exit 5
       ;;
    v) V_FLAG="$OPTARG";
       ;;
    d) JSON_TREE="$OPTARG";
       ;;
    i) LIMIT_SH="$OPTARG";
       ;;
    o) OLD_FLAG="-o"
       ;;
    F) FORCE="-f"
       ;;
    Q) QUOTE=
       ;;
    n) NOOP="-n"
       ;;
    l) LIST_CHANGE="-l"
       ;;
    L) LIST_NOCHANGE="-L"
       ;;
    \?) echo "$0: ERROR: invalid option: -$OPTARG" 1>&2
       exit 6
       ;;
    :) echo "$0: ERROR: option -$OPTARG requires an argument" 1>&2
       exit 6
       ;;
   *)
       ;;
    esac
done

# check args
#
shift $(( OPTIND - 1 ));
if [[ $# -ne 2 ]]; then
    echo "$0: ERROR: expected 2 arguments, found $#" 1>&2
    echo "$USAGE" 1>&2
    exit 6
fi
OLD_VER="$QUOTE$1$QUOTE"
NEW_VER="$QUOTE$2$QUOTE"
export OLD_VER NEW_VER

# firewall
#
if [[ ! -e $JSON_TREE ]]; then
    echo "$0: ERROR: test_dir does not exist: $JSON_TREE" 1>&2
    exit 4
fi
if [[ ! -d $JSON_TREE ]]; then
    echo "$0: ERROR: test_dir is not a directory: $JSON_TREE" 1>&2
    exit 4
fi
if [[ ! -r $JSON_TREE ]]; then
    echo "$0: ERROR: test_dir is not readable: $JSON_TREE" 1>&2
    exit 4
fi
if [[ ! -e $LIMIT_SH ]]; then
    echo "$0: ERROR: limit.sh does not exist: $LIMIT_SH" 1>&2
    exit 3
fi
if [[ ! -f $LIMIT_SH ]]; then
    echo "$0: ERROR: limit.sh is not a regular file: $LIMIT_SH" 1>&2
    exit 3
fi
if [[ ! -r $LIMIT_SH ]]; then
    echo "$0: ERROR: limit.sh is not a readable file: $LIMIT_SH" 1>&2
    exit 3
fi
RPL_CMD=$(type -P rpl)
if [[ -z "$RPL_CMD" ]]; then
    echo "$0: ERROR: rpl command not found" 1>&2
    exit 2
fi
if [[ $V_FLAG -ge 5 ]]; then
    echo "$0: debug[5]: V_FLAG: $V_FLAG" 1>&2
    echo "$0: debug[5]: test_dir: $JSON_TREE" 1>&2
    echo "$0: debug[5]: limit_ioccc.sh: $LIMIT_SH" 1>&2
    echo "$0: debug[5]: rpl: $RPL_CMD" 1>&2
    echo "$0: debug[5]: OLD_FLAG: <$OLD_FLAG>" 1>&2
    echo "$0: debug[5]: FORCE: <$FORCE>" 1>&2
    echo "$0: debug[5]: QUOTE: <$QUOTE>" 1>&2
    echo "$0: debug[5]: NOOP: <$NOOP>" 1>&2
    echo "$0: debug[5]: LIST_CHANGE: <$LIST_CHANGE>" 1>&2
    echo "$0: debug[5]: LIST_NOCHANGE: <$LIST_NOCHANGE>" 1>&2
    echo "$0: debug[5]: OLD_VER: <$OLD_VER>" 1>&2
    echo "$0: debug[5]: NEW_VER: <$NEW_VER>" 1>&2
fi

# count *.json files under test_dir
#
COUNT=$(find "$JSON_TREE" -type f -name '*.json' -print 2>/dev/null | wc -l)
export COUNT
if [[ $COUNT -le 0 ]]; then
    echo "$0: ERROR: no .json files found under test_dir: $JSON_TREE" 1>&2
    exit 7
fi
if [[ $V_FLAG -ge 3 ]]; then
    echo "$0: debug[3]: .json file count: $COUNT" 1>&2
fi

# check if a pattern is found in the limit_ioccc.sh file
#
if [[ -n $OLD_FLAG ]]; then
    if ! grep -F -q "$OLD_VER" "$LIMIT_SH" >/dev/null 2>&1; then
	if [[ -z $FORCE ]]; then
	    echo "$0: ERROR: old_ver: <$OLD_VER> not found in limit_ioccc.sh: $LIMIT_SH" 1>&2
	    exit 8
	else
	    echo "$0: Warning: old_ver: <$OLD_VER> not found in limit_ioccc.sh: $LIMIT_SH" 1>&2
	fi
    elif [[ $V_FLAG -ge 3 ]]; then
	echo "$0: debug[3]: old_ver: <$OLD_VER> found in limit_ioccc.sh: $LIMIT_SH" 1>&2
    fi
else
    if ! grep -F -q "$NEW_VER" "$LIMIT_SH" >/dev/null 2>&1; then
	if [[ -z $FORCE ]]; then
	    echo "$0: ERROR: new_ver: <$NEW_VER> not found in limit_ioccc.sh: $LIMIT_SH" 1>&2
	    exit 8
	else
	    echo "$0: Warning: new_ver: <$NEW_VER> not found in limit_ioccc.sh: $LIMIT_SH" 1>&2
	fi
    elif [[ $V_FLAG -ge 3 ]]; then
	echo "$0: debug[3]: new_ver: <$NEW_VER> found in limit_ioccc.sh: $LIMIT_SH" 1>&2
    fi
fi

# listing files
#
# case: -l and -L: list all .json files
#
if [[ -n $LIST_CHANGE && -n $LIST_NOCHANGE ]]; then
    find "$JSON_TREE" -type f -name '*.json' -print 2>/dev/null | sort -u

# case: -l: list .json files that WILL change
#
elif [[ -n $LIST_CHANGE ]]; then
# This warning is an example where shellcheck is incorrect. Yes '*.json' is a
# glob but --include '*.json' is not the actual pattern to search for.
# Specifically this option, if specified, says that only files matching the
# given filename pattern are searched.
#
# shellcheck disable=SC2063
     grep -F -R --include '*.json' -l -- "$OLD_VER" "$JSON_TREE" 2>/dev/null | sort -u

# case: -L: list .json files that WILL NOT change
#
elif [[ -n $LIST_NOCHANGE ]]; then
# This warning is an example where shellcheck is incorrect. Yes '*.json' is a
# glob but --include '*.json' is not the actual pattern to search for.
# Specifically this option, if specified, says that only files matching the
# given filename pattern are searched.
#
# shellcheck disable=SC2063
     grep -F -R --include '*.json' -L -- "$OLD_VER" "$JSON_TREE" 2>/dev/null | sort -u
fi

# replace string in files
#
if [[ $V_FLAG -ge 1 ]]; then
    echo "$0: debug[1]: rpl -q -x'.json' -R -- $OLD_VER $NEW_VER $JSON_TREE" 1>&2
fi
status=0
if [[ -z $NOOP ]]; then
    rpl -q -x'.json' -R -- "$OLD_VER" "$NEW_VER" "$JSON_TREE"
    status="$?"
    if [[ $status -ne 0 ]]; then
	echo "$0: ERROR: rpl non-zero exit status: $status" 1>&2
	exit 1
    fi
fi

# All Done!!! -- Jessica Noll, Age 2
#
exit 0
